api/src/services/critterbase-service.ts

```diff
@@ -0,0 +1,249 @@
+  children: { family_id: string; child_critter_id: string }[];
+}
+
+export interface IBulkCreate {
+  critters: ICritter[];
+  captures: ICapture[];
+  mortality: IMortality[];
+  locations: ILocation[];
+  markings: IMarking[];
+  measurement_quantitative: IQuantMeasurement[];
+  measurement_qualitative: IQualMeasurement[];
+  family: IFamilyPayload[];
+}
+
+export interface ICbSelectRows {
+  key: string;
+  id: string;
+  value: string;
+}
+
+const lookups = '/lookups';
+const xref = '/xref';
+const lookupsEnum = lookups + '/enum';
+const lookupsTaxons = lookups + '/taxons';
+export const CbRoutes = {
+  // lookups
+  ['region-envs']: `${lookups}/region-envs`,
+  ['region_nrs']: `${lookups}/region-nrs`,
+  wmus: `${lookups}/wmus`,
+  cods: `${lookups}/cods`,
+  ['marking-materials']: `${lookups}/marking-materials`,
+  ['marking-types']: `${lookups}/marking-types`,
+  ['collection-categories']: `${lookups}/collection-unit-categories`,
+  taxons: lookupsTaxons,
+  species: `${lookupsTaxons}/species`,
+  colours: `${lookups}/colours`,
+
+  // lookups/enum
+  sex: `${lookupsEnum}/sex`,
+  ['critter-status']: `${lookupsEnum}/critter-status`,
+  ['cause-of-death-confidence']: `${lookupsEnum}/cod-confidence`,
+  ['coordinate-uncertainty-unit']: `${lookupsEnum}/coordinate-uncertainty-unit`,
+  ['frequency-units']: `${lookupsEnum}/frequency-units`,
+  ['measurement-units']: `${lookupsEnum}/measurement-units`,
+
+  // xref
+  ['collection-units']: `${xref}/collection-units`,
+
+  // taxon xrefs
+  ['taxon-measurements']: `${xref}/taxon-measurements`,
+  ['taxon_qualitative_measurements']: `${xref}/taxon-qualitative-measurements`,
+  ['taxon-qualitative-measurement-options']: `${xref}/taxon-qualitative-measurement-options`,
+  ['taxon-quantitative-measurements']: `${xref}/taxon-quantitative-measurements`,
+  ['taxon-collection-categories']: `${xref}/taxon-collection-categories`,
+  ['taxon-marking-body-locations']: `${xref}/taxon-marking-body-locations`
+} as const;
+
+export type ICbRouteKey = keyof typeof CbRoutes;
+
+export const CRITTERBASE_API_HOST = process.env.CB_API_HOST || ``;
+const CRITTER_ENDPOINT = '/critters/';
+const BULK_ENDPOINT = '/bulk';
+const SIGNUP_ENDPOINT = '/signup';
+const FAMILY_ENDPOINT = '/family';
+
+export class CritterbaseService {
+  user: ICritterbaseUser;
+  keycloak: KeycloakService;
+
+  constructor(user: ICritterbaseUser) {
+    this.user = user;
+    this.keycloak = new KeycloakService();
+  }
+
+  async getToken(): Promise<string> {
+    const token = await this.keycloak.getKeycloakServiceToken();
+    return token;
+  }
+
+  getUserHeader(): { user: string } {
+    return { user: JSON.stringify(this.user) };
+  }
+
+  async makeGetRequest(endpoint: string, params: QueryParam[]) {
+    let url = `${CRITTERBASE_API_HOST}${endpoint}`;
+    for (let i = 0; i < params.length; i++) {
+      url += (i === 0 ? '?' : '&') + params[i].key + '=' + params[i].value;
```

Comment:
Apart from using the built-in `String.join` method for joining these together, Javascript has a way to build query URLS:

```javascript
// Create a new URLSearchParams object
const params = new URLSearchParams();

// Add parameters to the object
params.append('a', 1);
params.append('b', 2);
params.append('c', 3);

// Construct the URL with the query parameters
const url = `https://critters.com/?${params.toString()}`;

console.log(url);
```